src/Microsoft.AspNetCore.Mvc.Testing/WebApplicationFactory.cs

```diff
@@ -0,0 +1,313 @@
+using Microsoft.AspNetCore.TestHost;
+using Microsoft.Extensions.DependencyModel;
+
+namespace Microsoft.AspNetCore.Mvc.Testing
+{
+    /// <summary>
+    /// Factory for bootstrapping an application in memory for functional end to end tests.
+    /// </summary>
+    /// <typeparam name="TEntryPoint">A type in the entry point assembly of the application.
+    /// Typically the Startup or Program classes can be used.</typeparam>
+    public class WebApplicationFactory<TEntryPoint> : IDisposable where TEntryPoint : class
+    {
+        private TestServer _server;
+        private IWebHostBuilder _builder;
+        private IList<HttpClient> _clients = new List<HttpClient>();
+
+        /// <summary>
+        /// <para>
+        /// Creates a TestServer instance using the MVC application defined by <typeparamref name="TEntryPoint"/>.
+        /// The <see cref="WebApplicationFactory{TEntryPoint}"/> will find the entry point class of <typeparamref name="TEntryPoint"/>
+        /// assembly and initialize the application by calling <c>IWebHostBuilder CreateWebHostBuilder(string [] args)</c>
+        /// on <typeparamref name="TEntryPoint"/>.
+        /// </para>
+        /// <para>
+        /// This constructor will infer the application content root path by searching for an
+        /// <see cref="WebApplicationFactoryContentRootAttribute"/> on the assembly containing the functional tests with
+        /// a key equal to the <typeparamref name="TEntryPoint"/> assembly <see cref="Assembly.FullName"/>.
+        /// In case an attribute with the right key can't be found, <see cref="WebApplicationFactory{TEntryPoint}"/>
+        /// will fall back to searching for a solution file (*.sln) and then appending <typeparamref name="TEntryPoint"/> asembly name
+        /// to the solution directory. The application root directory will be used to discover views and content files.
+        /// </para>
+        /// <para>
+        /// The application assemblies will be loaded from the dependency context of the assembly containing
+        /// <typeparamref name = "TEntryPoint" />.This means that project dependencies of the assembly containing
+        /// <typeparamref name = "TEntryPoint" /> will be loaded as application assemblies.
+        /// </para>
+        /// </summary>
+        public WebApplicationFactory()
+        {
+        }
+
+        /// <summary>
+        /// Gets the <see cref="TestServer"/> created by this <see cref="WebApplicationFactory{TEntryPoint}"/>.
+        /// </summary>
+        public TestServer Server => _server;
+
+        /// <summary>
+        /// The <see cref="IWebHostBuilder"/> used to create the <see cref="TestServer"/>.
+        /// </summary>
+        public IWebHostBuilder WebHostBuilder
+        {
+            get
+            {
+                if (_server != null)
+                {
+                    throw new InvalidOperationException($"The '{nameof(TestServer)}' instance was already built.");
```

Comment:
Property getters really, really, *really* shouldn't throw. I would rather remove this exception and instead make a note in the doc comment that this property can't be used once the server is created. Alternatively, coming up with a different design for this would also work. That is, don't have a single object (this type) that is needed both for construction and for running.